src/NuGet.Protocol.Core.v2/DownloadResourceV2.cs

```diff
@@ -66,5 +69,211 @@ public DownloadResourceV2(V2Resource resource)
+            var dataServiceRepo = repository as DataServicePackageRepository;
+
+            if (dataServiceRepo != null)
+            {
+                // Clone the repo to allow for concurrent calls
+                var sourceUri = new Uri(dataServiceRepo.Source);
+                dataServiceRepo = new DataServicePackageRepository(sourceUri);
+
+                var package = dataServiceRepo.FindPackage(identity.Id, version);
+                token.ThrowIfCancellationRequested();
+
+                // For online sources get the url and retrieve it with cancel support
+                var dataServicePackage = package as DataServicePackage;
+                var url = dataServicePackage.DownloadUrl;
+
+                var downloadedPackage = DownloadToMachineCache(
+                    MachineCache.Default,
+                    identity,
+                    dataServiceRepo,
+                    url,
+                    token);
+
+                if (downloadedPackage != null)
+                {
+                    return new DownloadResourceResult(downloadedPackage.GetStream());
+                }
+            }
+            else
+            {
+                var package = repository.FindPackage(identity.Id, version);
+
+                // Use a folder reader for unzipped repos
+                if (repository is UnzippedPackageRepository)
+                {
+                    var packagePath = Path.Combine(repository.Source, identity.Id + "." + version);
+                    var directoryInfo = new DirectoryInfo(packagePath);
+                    if (directoryInfo.Exists)
+                    {
+                        return new DownloadResourceResult(
+                            package.GetStream(),
+                            new PackageFolderReader(directoryInfo));
+                    }
+                }
+
+                return new DownloadResourceResult(package.GetStream());
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// True if the given package matches hash
+        /// </summary>
+        private static bool MatchPackageHash(IPackage package, string hash)
+        {
+            var hashProvider = new CryptoHashProvider("SHA512");
+
+            return package != null && package.GetHash(hashProvider).Equals(hash, StringComparison.OrdinalIgnoreCase);
+        }
+
+        private static IPackage DownloadToMachineCache(
+            IPackageCacheRepository cacheRepository,
+            PackageIdentity package,
+            DataServicePackageRepository dataServiceRepo,
+            Uri downloadUri,
+            CancellationToken token)
+        {
+            var packageName = new PackageNameWrapper(package);
+            var version = SemanticVersion.Parse(package.Version.ToString());
+            IPackage newPackage = null;
+
+            FileInfo tmpFile = null;
+            FileStream tmpFileStream = null;
+
+            var downloadClient = new HttpClient(downloadUri)
```

Comment:
I would add a comment that this is not the HttpClient from Microsoft.Net.Http package (or System.Net.Http.dll)